File: apps/haskell/test/FizzBuzz/TypeSpec.hs
module FizzBuzz.TypeSpec (spec) where

import FizzBuzz.Type
import Test.Hspec

spec :: Spec
spec = do
  describe "Type01" $ do
    it "1 を渡すと '1' を返す" $
      generate Type01 1 `shouldBe` "1"

    it "3 の倍数を渡すと 'Fizz' を返す" $
      generate Type01 3 `shouldBe` "Fizz"

    it "5 の倍数を渡すと 'Buzz' を返す" $
      generate Type01 5 `shouldBe` "Buzz"

    it "15 の倍数を渡すと 'FizzBuzz' を返す" $
      generate Type01 15 `shouldBe` "FizzBuzz"

  describe "Type02" $ do
    it "数値を文字列に変換する" $
      generate Type02 1 `shouldBe` "1"

  describe "Type03" $ do
    it "1 を渡すと '1' を返す" $
      generate Type03 1 `shouldBe` "1"

    it "3 を渡すと 'Fizz' を返す" $
      generate Type03 3 `shouldBe` "Fizz"

    it "15 の倍数で 'FizzBuzz' を返す" $
      generate Type03 15 `shouldBe` "FizzBuzz"

  describe "createType" $ do
    it "1 を渡すと Type01 を返す" $
      createType 1 `shouldBe` Right Type01

    it "2 を渡すと Type02 を返す" $
      createType 2 `shouldBe` Right Type02

    it "3 を渡すと Type03 を返す" $
      createType 3 `shouldBe` Right Type03

    it "未定義のタイプはエラーを返す" $
      createType 4 `shouldBe` Left "未定義のタイプです"

File: apps/scala/src/test/scala/fizzbuzz/TypeSpec.scala
package fizzbuzz

import fizzbuzz.domain.FizzBuzzType
import org.scalatest.funsuite.AnyFunSuite

class TypeSpec extends AnyFunSuite:
  test("Type01: 3 の倍数で Fizz を返す") {
    assert(FizzBuzzType.Type01.generate(3) === "Fizz")
  }

  test("Type02: 15 を渡すと 15 を返す") {
    assert(FizzBuzzType.Type02.generate(15) === "15")
  }

  test("Type03: 15 の倍数で FizzBuzz を返す") {
    assert(FizzBuzzType.Type03.generate(15) === "FizzBuzz")
  }

  test("create: タイプ 1 を生成できる") {
    assert(FizzBuzzType.create(1) === FizzBuzzType.Type01)
  }

  test("create: タイプ 2 を生成できる") {
    assert(FizzBuzzType.create(2) === FizzBuzzType.Type02)
  }

  test("create: タイプ 3 を生成できる") {
    assert(FizzBuzzType.create(3) === FizzBuzzType.Type03)
  }

  test("create: 未定義のタイプで例外が発生する") {
    assertThrows[IllegalArgumentException] {
      FizzBuzzType.create(4)
    }
  }
